File: Week 2/cpp/6.fahrenheit_to_celsius_table.cpp
/*
 * WEEK 2 - C++ CONTROL FLOW
 * Topic: Fahrenheit -> Celsius Table
 * File: 6.fahrenheit_to_celsius_table.cpp
 *
 * PROBLEM:
 *  Given start S, end E, step W, print (F, C) pairs where
 *  C = (F - 32) * 5 / 9, for F = S, S+W, ..., <= E.
 *
 * CONCEPT:
 *  Integer arithmetic truncates the Celsius value -- exact same behaviour
 *  as Java. For a precise float result use `(F - 32) * 5.0 / 9.0`.
 *
 * KEY POINTS:
 *  - int / int truncates toward zero.
 *  - `(int) ((F - 32) * 5.0 / 9.0)` is the precise -> truncated path.
 *  - Loop pattern: while (S <= E) { ... S += W; }
 *
 * DRY RUN:
 *  S=32, E=212, W=20
 *    F=32  -> C=0
 *    F=52  -> C=11
 *    ...
 *    F=212 -> C=100
 *
 * COMPLEXITY: O((E - S)/W).
 */

#include <iostream>

int main() {
    int S, E, W;
    if (!(std::cin >> S >> E >> W)) {
        // Demo defaults if no input piped
        S = 32; E = 212; W = 20;
        std::cout << "(demo S=32 E=212 W=20)\n";
    }
    if (W <= 0) {
        std::cerr << "step W must be > 0\n";
        return 1;
    }

    for (int F = S; F <= E; F += W) {
        int  C_int = (F - 32) * 5 / 9;            // matches Java
        double C_d = (F - 32) * 5.0 / 9.0;        // precise
        std::cout << F << " " << C_int
                  << "   (precise: " << C_d << ")\n";
    }
    return 0;
}

/*
 * NOTES:
 *  - C++ truncates int/int just like Java (toward zero).
 *  - Always pick the divisor's type carefully -- use `5.0/9.0` if you want a double.
 *  - Range-based for cannot count by W directly; we use a classic `for`.
 */

File: Week 2/python/6.fahrenheit_to_celsius_table.py
"""
WEEK 2 - PYTHON CONTROL FLOW
Topic: Fahrenheit -> Celsius Table
File: 6.fahrenheit_to_celsius_table.py

PROBLEM:
Given START fahrenheit (S), END fahrenheit (E), and STEP W, print every
(F, C) pair where C = (F - 32) * 5 / 9, F runs from S to E in steps of W.

CONCEPT:
A while loop walks F from S to E, increments by W each iteration. The Java
example uses integer arithmetic (so C is truncated). Python's `/` is float
division by default; we use `//` to mimic the Java truncation exactly.

KEY POINTS:
 - `(F - 32) * 5 // 9` -> integer (Java-style) Celsius.
 - `(F - 32) * 5 / 9`  -> precise float Celsius.
 - Loop pattern: initialise, test, body, update.

SYNTAX:
 while s <= e:
     c = (s - 32) * 5 // 9
     print(s, c)
     s += w

DRY RUN:
 S=32, E=212, W=20
   F=32  -> C = 0
   F=52  -> C = 11   (Java int trunc; precise is 11.111...)
   F=72  -> C = 22
   ...
   F=212 -> C = 100

COMPLEXITY: O((E - S) / W) iterations.
"""

import sys


def print_table(s: int, e: int, w: int) -> None:
    if w <= 0:
        raise ValueError("step W must be positive")
    while s <= e:
        c_int = (s - 32) * 5 // 9            # matches Java
        c_flt = (s - 32) * 5 / 9              # precise
        print(f"{s} {c_int}   (precise: {c_flt:.4f})")
        s += w


def main() -> None:
    if sys.stdin.isatty():
        # Demo defaults
        print("(demo) S=32, E=212, W=20")
        print_table(32, 212, 20)
        return

    tokens = sys.stdin.read().split()
    s, e, w = int(tokens[0]), int(tokens[1]), int(tokens[2])
    print_table(s, e, w)


if __name__ == "__main__":
    main()


# NOTES:
# - Java's `int` arithmetic truncates the conversion result. Use `//` in Python
#   to match exactly.
# - `range(s, e + 1, w)` is the idiomatic Python alternative to a manual while loop:
#       for f in range(s, e + 1, w): ...
# - For floating-point loops, prefer fixed integer counters then multiply by a step,
#   to avoid accumulated float errors.
